src/main/java/com/xceptance/xlt/mastercontroller/AgentControllerStatusUpdater.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2005-2022 Xceptance Software Technologies GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.xceptance.xlt.mastercontroller;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.xceptance.common.util.concurrent.DaemonThreadFactory;
+import com.xceptance.xlt.agentcontroller.AgentController;
+import com.xceptance.xlt.agentcontroller.AgentControllerStatus;
+import com.xceptance.xlt.agentcontroller.AgentStatus;
+import com.xceptance.xlt.util.StatusUtils;
+
+/**
+ * Retrieves the status from all agent controllers and stores that internally.
+ * <p>
+ * Implementation note: The natural and simpler solution would be to use a scheduled executor service to run the update
+ * tasks periodically. However, the underlying scheduled thread pool is always fixed size. In XLT with its default
+ * settings, the default thread pool is dynamic, though. To preserve this behavior, we combine a scheduled executor
+ * service with a regular executor service. The former is only responsible for the timing and simply delegates the hard
+ * work to the latter. Hence a single-thread scheduled executor should suffice.
+ */
+public class AgentControllerStatusUpdater
+{
+    private static final Logger LOG = LoggerFactory.getLogger(AgentControllerStatusUpdater.class);
+
+    /**
+     * The agent controller status objects keyed by agent controller name.
+     */
+    private final Map<String, AgentControllerStatusInfo> statusMap;
+
+    /**
+     * The agent controller status objects keyed by agent controller name as an unmodifiable map.
+     */
+    private final Map<String, AgentControllerStatusInfo> unmodifiableStatusMap;
```

Comment:
We could create the unmodifiable view also in `getAgentControllerStatusMap`, but then it would be created again and again, each time the current agent status is requested by the user interface. Hence the instance field.

Reply:
Creating short-living objects should not be an issue. Thus, I don't think that this optimization is really necessary.
Anyway, thanks for clarification.